src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,156 @@
+            }
+            if (!dataFile.exists()) {
+                dataFile.createNewFile();
+            }
+        } catch (IOException e) {
+            System.out.println("Unable to create file!!");
+        }
+        return dataFile;
+    }
+
+    /**
+     * Retrieves the tasks stored in the ./data/duke.txt file.
+     * Converts each task string into their corresponding Task object.
+     * Tasks are then added to an ArrayList.
+     *
+     * @return an ArrayList of Task objects.
+     */
+    public ArrayList<Task> loadTasks() {
+        File dataFile = new File(this.path);
+        if (!dataFile.exists()) {
+            dataFile = createDataFile();
+        }
+        ArrayList<Task> list = new ArrayList<>();
+
+        try {
+            Scanner sc = new Scanner(dataFile);
+
+            while (sc.hasNextLine()) {
+                String task = sc.nextLine();
+                if (!task.isBlank()) {
+                    // | is a special symbol
+                    String[] taskDetails = task.split(" " + "\\|" + " ");
+                    String type = taskDetails[0];
+                    int status = Integer.parseInt(taskDetails[1]);
+                    String desc = taskDetails[2];
+
+                    switch (type) {
+                    case "T":
+                        ToDo toDo = new ToDo(status, desc);
+                        list.add(toDo);
+                        break;
+                    case "D":
+                        LocalDateTime date = convertToDateTime(taskDetails[3]);
+                        Deadline deadline = new Deadline(status, desc, date);
+                        list.add(deadline);
+                        break;
+                    case "E":
+                        LocalDateTime start = convertToDateTime(taskDetails[3]);
+                        LocalDateTime end = convertToDateTime(taskDetails[4]);
+                        try {
+                            Event event = new Event(status, desc, start, end);
+                            list.add(event);
+                        } catch (InvalidStartEndException e) {
+                            ui.showError(e.getMessage());
+                        }
+                        break;
+                    }
+                }
+            }
+            sc.close();
+        } catch (IOException e) {
+            ui.showError(e.getMessage());
+        }
+        return list;
+    }
+
+    /**
+     * Writes the tasks in the given list to the data file.
+     * This overwrites the existing data in the file.
+     *
+     * @param list list of Task objects.
+     */
+    public void updateFile(ArrayList<Task> list) {
+        try {
+            //check if file exists, else create
+            File dataFile = new File("./data/duke.txt");
+            if (!dataFile.exists()) {
+                dataFile = createDataFile();
+            }
+
+            //create a FileWriter object to write to file. Note that this overwrites the existing data!
```

Comment:
I like this warning for i have fallen for this so many times D:
I still don't know how FileWriters work